### ISSUE TITLE
Inclusion of null values in code generation

### ISSUE THREAD

**Issue by walsha2** [tazatechnology/openapi_spec#16]
> I have went ahead and added this as another test. That is a pretty comprehensive spec that should be supported out of the box, IMO. It actually did uncover some minor issues that I resolved and have an impact on this issue's feature request. So that's good.

Pretty good progress! 

> The model enum is reflected as a doc string at the moment and type hint denotes String - just as defined in the spec.

This is already a good improvement.

> I didn't have any time to test against the OpenAI API, but the code looks good.

I've played a bit with the completions endpoint and these are the findings:

#### Server URL

Currently, the generator only takes the schema and host from the defined server. It ignores any defined path.

So for:
```yaml
servers:
  - url: https://api.openai.com/v1
```

The generator generates:
```dart
final r = await _request(
  host: 'https://api.openai.com',
  path: '/completions',
  ...
);
```
(the `v1` is not included)

Maybe we should rename `host` to `baseUrl` while fixing it.

**Update:** I've issued a fix in https://github.com/tazatechnology/openapi_spec/pull/15

#### `CreateCompletionRequest`

The generated class looks good. The only issue that I've encountered is that `nulls` are also serialized and OpenAI doesn't seem to like it.

Eg.

Calling:
```dart
final res = await client.createCompletion(
  request: CreateCompletionRequest(
    model: 'gpt-3.5-turbo-instruct',
    prompt: 'Repeat: This is a test',
  ),
);
```

Generates:
```json
{
  "model": "gpt-3.5-turbo-instruct",
  "prompt": "Repeat: This is a test",
  "best_of": 1,
  "echo": false,
  "frequency_penalty": 0.0,
  "logit_bias": null,
  "logprobs": null,
  "max_tokens": 16,
  "n": 1,
  "presence_penalty": 0.0,
  "stop": null,
  "stream": false,
  "suffix": null,
  "temperature": 1.0,
  "top_p": 1.0,
  "user": null
}
```

But OpenAI replies with the following error:
```
OpenAIClientException({
  "uri": "https://api.openai.com/v1/completions",
  "method": "POST",
  "code": 400,
  "message": "Unsuccessful response",
  "body": {
    "error": {
      "message": "None is not of type 'object' - 'logit_bias'",
      "type": "invalid_request_error",
      "param": null,
      "code": null
    }
  }
})
```

I'm not sure what's the best default behaviour for the generator, should it include `nulls` or not? 

I've checked [`dart-dio`](https://openapi-generator.tech/docs/generators/dart-dio) generator and it doesn't include them (it generates: `@JsonKey(name: r'logit_bias', required: false, includeIfNull: false)`).

Anyway, it's not a blocking issue as you can always configure `include_if_null: false` in the `build.yaml` file.

#### `CreateCompletionResponse`

The `choices` array is generated as `List<Map<String, dynamic>>` instead of a List of the inner object. 

[`dart-dio`](https://openapi-generator.tech/docs/generators/dart-dio) generator creates a `final List<CreateCompletionResponseChoicesInner> choices` by default. So they concatenate the schema name (`CreateCompletionResponse`), plus the field (`choices`) plus `Inner` to create the class name of the inner object.

_Originally posted by @davidmigloz in https://github.com/tazatechnology/openapi_spec/issues/13#issuecomment-1770441157_

**Comment by walsha2** [comment]
> Anyway, it's not a blocking issue as you can always configure include_if_null: false in the build.yaml file.

@davidmigloz moving to a new issue for this specific item. Using `build.yaml` is how I typically resolve this issue (as opposed to adding more verbose generated code).

> I'm not sure what's the best default behaviour for the generator, should it include nulls or not?

This is the real question. If we believe the default should be to not include it, then **we should** add `includeIfNull: false` to the `JsonKey` for all nullable variables. I need to do a little bit of research on a best practice, but my initial hunch is that they should be removed by default. 

If we really want we can even add it as a generator option:

```dart
spec.generate(
  schemaOptions: SchemaGeneratorOptions(
    enabled: true,
    includeIfNull: false // Default is false
  ),
);
```

Let me know your thoughts.

**Comment by walsha2** [comment]
@davidmigloz I thought about this a bit more after working a PR, and it is a bit hairy when you dig into the details. There is not enough information in the OpenAPI spec to know if `include_if_null` should be true/false and in turn it should be left to the user. The default `json_serializable` build configuration is linked below. You will see the default is `include_if_null: true`. 

https://pub.dev/packages/json_serializable#build-configuration

Overriding it at the schema code generation will create unexpected conflicts and essentially hard code the `include_if_null` setting in the schema generation `JsonKey` as opposed to allow the user to control it in the `build.yaml` code. 

The purpose of `JsonKey` in the schema generator is primarily used to ensure that names are handled, one of the more annoying aspects of creating these schemas by hand. All other build options are left to the user to control via the `build.yaml` configuration as they see fit.

Not to mention, it adds a ton of code generation changes in each file when it is likely best handled by a single `include_if_null: false` line in the `build.yaml` file. 

-----------

TLDR: I think we should keep things as is and allow the user to control this at `build.yaml`. I have the code to satisfy this issue ready to go via `SchemaGeneratorOptions`, but after writing it, am not so sure that is the best solution due to the comments above. Feedback welcome - else, will close this issue.

**Comment by davidmigloz** [comment]
Thanks for all the details!

One option could be to add `includeIfNull: false` to fields that are `nullable: true` and `required: false`. 
So if you still want to serialize the null value you just need to make that field `required`. Would that work?

**Comment by walsha2** [comment]
> One option could be to add includeIfNull: false to fields that are nullable: true and required: false.

Wouldnt that still conflict with this primary issue, the fact that it is now hardcoded in the generated schema:

> Overriding it at the schema code generation will ... hard code the `include_if_null` setting in the schema generation `JsonKey` as opposed to allow the user to control it in the build.yaml code.

**Comment by davidmigloz** [comment]
> Wouldnt that still conflict with this primary issue, the fact that it is now hardcoded in the generated schema
Isn't that a desired feature in a spec?

> There is not enough information in the OpenAPI spec to know if include_if_null should be true/false and in turn it should be left to the user

In which cases we don't have enough information?

- If the server expects all the fields to be serialized (even the null ones), then all fields should be marked as required in the spec (including the nullable ones)
- If the server doesn't expect nullable fields to be serialized, then the nullable fields should not be marked as required

**Comment by walsha2** [comment]
> Isn't that a desired feature in a spec?

Yea I wont argue with that. Regarding your example above, the issue is "should be marked" or "should not be marked"... although they _should_ (good practice), that is not always the case.

------

> One option could be to add includeIfNull: false to fields that are nullable: true and required: false.
So if you still want to serialize the null value you just need to make that field required

In any case, I think this is a reasonable middle ground and I can get that in! Thanks for the discussion @davidmigloz
